### PR TITLE
use EndpointSlices by default in 1.21

### PR DESCRIPTION
**Pull request by stevenctl**
Not sure if we have other examples changing default behavior conditionally on the k8s version, this seems like it would wor.

### CODE REVIEW

**Comment by stevenctl** [comment]
Should this have a release note? I would think it's sort of an implementation detail of pilot.

File: tests/integration/pilot/endpointslice/endpointslice_test.go [hunk 1]
```diff
@@ -41,11 +43,13 @@ func TestMain(m *testing.M) {
...
 		Setup(istio.Setup(&i, func(t resource.Context, cfg *istio.Config) {
-			cfg.ControlPlaneValues = `
+			cfg.ControlPlaneValues = fmt.Sprintf(`
 values:
   pilot:
     env:
-      PILOT_USE_ENDPOINT_SLICE: "true"`
+      PILOT_USE_ENDPOINT_SLICE: "true"`,
+      			// for k8s 1.21+, this suite should test disabling EndpointSlice mode
+				kubelib.IsLessThanVersion(t.Clusters().Kube().Default(), 21))
```
Comment (howardjohn) [inline tests/integration/pilot/endpointslice/endpointslice_test.go:52]: the sprintf doesn't seem to have any `%s`?

**Comment by howardjohn** [comment]
Blocker: https://github.com/istio/istio/issues/34995

**Comment by howardjohn** [comment]
> Should this have a release note? I would think it's sort of an implementation detail of pilot.

IMO yes. It is an implementation detail but a big enough one users should know about it for debugging, etc.

Also note users may manually create EndpointSlices that are currently ignored by Istiod, which will now not be ignored. So it can be user facing in edge cases

**Review by hzxuzhonghu — APPROVED** [review]

**Review by howardjohn — CHANGES_REQUESTED** [review]
This LGTM but would prefer https://github.com/istio/istio/issues/34995 to be fixed first. While its a relatively rare edge case, the impact is extremely bad and it is a regression

**Comment by stevenctl** [comment]
With #35990 do we want to squeeze this into 1.12?

**Comment by howardjohn** [comment]
This seems like a pretty large change to get this late into the release. I
think its ideal to get it in now for master so we get 4 months of soak time
before shipping?

**Comment by ramaraochavali** [comment]
+1 for master

**Review by howardjohn — APPROVED** [review]

**Pull request merged**